feat(linter): add eslint/no-unused-vars (⭐ attempt 3.2)

DonIsaac opened the pull request:
> Re-creation of #4427 due to rebasing issues. Original attempt: #642
-----

Third time's the charm?

Each time I attempt this rule, I find a bunch of bugs in `Semantic`, and I expect this attempt to be no different. Expect sidecar issues+PRs stemming from this PR here.

## Not Supported
These are cases supported in the original eslint rule, but that I'm intentionally deciding not to support
- export comments in scripts
  ```js
  /* exported a */ var a;
  ```
- global comments
  ```js
  /* global a */ var a;
   ```

## Behavior Changes
These are intentional deviations from the original rule's behavior:
- logical re-assignments are not considered usages
  ```js
  // passes in eslint/no-unused-vars, fails in this implementation
  let a = 0; a ||= 1;
  let b = 0; b &&= 2;
  let c = undefined; c ??= []
  ```

## Known Limitations
- Lint rules do not have babel or tsconfig information, meaning we can't determine if `React` imports are being used or not. The relevant tsconfig settings here are `jsx`, `jsxPragma`, and `jsxFragmentName`. To accommodate this, all imports to symbols named `React` or `h` are ignored in JSX files.
- References to symbols used in JSDoc `{@link}` tags are not created, so symbols that are only used in doc comments will be reported as unused. See: #4443
- `.vue` files are skipped completely, since variables can be used in templates in ways we cannot detect
  > note: `.d.ts` files are skipped as well.

## Todo
- [x] Skip unused TS enum members on used enums
- [x] Skip unused parameters followed by used variables in object/array spreads
- [x] Re-assignments to array/object spreads do not respect `destructuredArrayIgnorePattern` (related to: https://github.com/oxc-project/oxc/issues/4435)
- [x] #4493
- [x] References inside a nested scope are not considered usages (#4447)
- [x] Port over typescript-eslint test cases _(wip, they've been copied and I'm slowly enabling them)_
- [x] Handle constructor properties
  ```ts
  class Foo {
    constructor(public a) {} // `a` should be allowed
  }
  ```
- [x] Read references in sequence expressions (that are not in the last position) should not count as a usage
  ```js
  let a = 0; let b = (a++, 0); console.log(b)
  ```
  > Honestly, is anyone even writing code like this?
- [x] function overload signatures should not be reported
- [x] Named functions returned from other functions get incorrectly reported as unused (found by @camc314)
  ```js
  function foo() {
    return function bar() { }
  }
  Foo()()
  ```
- [x] false positive for TS modules within ambient modules
  ```ts
  declare global {
    // incorrectly marked as unused
    namespace jest { }
  }
  ```
  

## Blockers
- https://github.com/oxc-project/oxc/issues/4436
- https://github.com/oxc-project/oxc/issues/4437
- #4446 
- #4447
- #4494
- #4495

## Non-Blocking Issues
- #4443
- #4475 (prevents checks on exported symbols from namespaces)

Dunqing commented:
Nice performance! I'll review this PR this weekend.

camc314 commented:
i get a false positive with:

```
function test() {
    return function Baz() {
        return 'Baz';
    };
}

test()();
```

output:

```
  × eslint(no-unused-vars): Function 'Baz' is declared but never used.
   ╭─[./test.ts:2:21]
 1 │ function test() {
 2 │     return function Baz() {
   ·                     ─┬─
   ·                      ╰── 'Baz' is declared here
 3 │         return 'Baz';
   ╰────
  help: Consider removing this declaration.

Finished in 7ms on 1 file with 1 rules using 12 threads.
Found 0 warnings and 1 error.
```

DonIsaac commented:
@camc314 thank you so much for the feedback! I'll add those cases and fix it up.

camc314 commented:
this one looks like another false positive:

Input:

```
Promise.withResolvers ||
    (Promise.withResolvers = function withResolvers<T>() {
        let resolve!: (value: T | PromiseLike<T>) => void;
        let reject!: (reason: unknown) => void;

        const promise = new this<T>((promiseResolve, promiseReject) => {
            resolve = promiseResolve;
            reject = promiseReject;
        });

        return {
            resolve,
            reject,
            promise,
        };
    });
```

output:
```

  × eslint(no-unused-vars): Function 'withResolvers' is declared but never used.
   ╭─[./test.ts:2:39]
 1 │ Promise.withResolvers ||
 2 │     (Promise.withResolvers = function withResolvers<T>() {
   ·                                       ──────┬──────
   ·                                             ╰── 'withResolvers' is declared here
 3 │         let resolve!: (value: T | PromiseLike<T>) => void;
   ╰────
  help: Consider removing this declaration.

Finished in 8ms on 1 file with 1 rules using 12 threads.
Found 0 warnings and 1 error.
```

DonIsaac commented:
@camc314 all above cases should be fixed now.

camc314 commented:
> @camc314 all above cases should be fixed now.

awesome, this is great work!

i still get some false positives, but only ever in minified code, so i think those are fine (i can't see why you would write code like this 🙂)?

looks like both have a related cause (sequence expressions)

these were the final two issues where i saw errors:

input:
```ts
class Test {
    async updateContextGroup(t, i, s = !0) {
        s ? await this.leave(i) : await this.join(t, i), false;
    }
}

new Test();
```

output:
```

  × eslint(no-unused-vars): Parameter 's' is declared but never used.
   ╭─[./test.ts:2:36]
 1 │ class Test {
 2 │     async updateContextGroup(t, i, s = !0) {
   ·                                    ┬
   ·                                    ╰── 's' is declared here
 3 │         s ? await this.leaveContextGroup(i) : await this.joinContextGroup(t, i), false;
   ╰────
  help: Consider removing this parameter.

Finished in 2ms on 1 file with 1 rules using 12 threads.
Found 0 warnings and 1 error.
```

input:
```ts
const l = 'undefined' != typeof window ? window : {},
    O = async () => {
        P(), o && (o.map((e) => e()), (l['s-rc'] = void 0));
    },
    P = () => {};

O();
```

output:
```

  × eslint(no-unused-vars): Variable 'P' is declared but never used.
   ╭─[./test.ts:5:5]
 4 │     },
 5 │     P = () => {};
   ·     ┬
   ·     ╰── 'P' is declared here
 6 │ 
   ╰────
  help: Consider removing this declaration.

Finished in 11ms on 1 file with 1 rules using 12 threads.
Found 0 warnings and 1 error.
```

camc314 commented on crates/oxc_linter/src/snapshots/no_unused_vars@oxc-classes.snap:
```diff
@@ -0,0 +1,28 @@
+---
+  ⚠ eslint(no-unused-vars): Parameter 'a' is declared but never used.
+   ╭─[no_unused_vars.tsx:1:32]
+ 1 │ export class Foo { constructor(a: number) {} }
+   ·                                ────┬────
+   ·                                    ╰── 'a' is declared here
```
this is very much a nit, but would be great if this could only report on the `a` (identifier) rather than the whole arg

DonIsaac replied:
I agree. Spans for symbols currently cover both the binding and the type annotation. I'll make a separate PR updating this. CC: @Dunqing  @overlookmotel

DonIsaac replied:
Nevermind, this is intentional behavior for the parser. See #2624. Is it worth trying to subtract out type annotations from the reported span?

Dunqing replied:
I have looked at `TypeScript`, which doesn't cover the type annotation span.

camc314 commented on crates/oxc_linter/src/snapshots/no_unused_vars@typescript-eslint-tsx.snap:
should this be empty?

DonIsaac replied:
Yes, I haven't ported the TSX failure cases yet.

camc314 commented on crates/oxc_linter/src/snapshots/no_unused_vars@eslint.snap:
```diff
@@ -0,0 +1,1526 @@
+
+  ⚠ eslint(no-unused-vars): Variable 'err' is caught but never used.
+   ╭─[no_unused_vars.tsx:1:12]
+ 1 │ try{}catch(err){};
+   ·            ─┬─
+   ·             ╰── 'err' is declared here
+   ╰────
+  help: Consider handling this error.
+
+  ⚠ eslint(no-unused-vars): Variable 'err' is caught but never used.
+   ╭─[no_unused_vars.tsx:1:12]
+ 1 │ try{}catch(err){};
+   ·            ─┬─
+   ·             ╰── 'err' is declared here
+   ╰────
+  help: Consider handling this error.
+
+  ⚠ eslint(no-unused-vars): Variable 'err' is caught but never used.
+   ╭─[no_unused_vars.tsx:1:12]
+ 1 │ try{}catch(err){};
+   ·            ─┬─
+   ·             ╰── 'err' is declared here
+   ╰────
+  help: Consider handling this error.
+
+  ⚠ eslint(no-unused-vars): Variable 'err' is caught but never used.
+   ╭─[no_unused_vars.tsx:1:12]
+ 1 │ try{}catch(err){};
+   ·            ─┬─
+   ·             ╰── 'err' is declared here
+   ╰────
+  help: Consider handling this error.
+
+  ⚠ eslint(no-unused-vars): Variable 'err' is caught but never used.
+   ╭─[no_unused_vars.tsx:1:12]
+ 1 │ try{}catch(err){};
+   ·            ─┬─
+   ·             ╰── 'err' is declared here
+   ╰────
+  help: Consider handling this error.
```
this reports a bunch of errors on the same symbol?

camc314 replied:
ah they are testing different configurations, all good!

DonIsaac commented:
> ```ts
> class Test {
>     async updateContextGroup(t, i, s = !0) {
>         s ? await this.leave(i) : await this.join(t, i), false;
>     }
> }
> 
> new Test();
> ```

I added support for the first case, but not the second. In non-function-call cases, or function calls with no side effects, that's not used. Its _possible_ that it could be used, but it's so pathological that I gave up trying to add it...

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/mod.rs:
```diff
@@ -0,0 +1,338 @@
+        };
+    }
+
+    // TODO: use when [#4446](https://github.com/oxc-project/oxc/issues/4446)
+    // is closed.
+    #[allow(dead_code)]
```
#4446 closed as not planned. This should be deleted.

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/symbol.rs:
```diff
@@ -0,0 +1,225 @@
+    ///
+    /// NOTE: does not support CJS right now.
+    pub fn is_exported(&self) -> bool {
+        let is_in_exportable_scope = self.is_root() || self.is_in_ts_namespace();
+        (is_in_exportable_scope
+            && (self.flags.contains(SymbolFlags::Export)
+                || self.semantic.module_record().exported_bindings.contains_key(self.name())))
+            || self.in_export_node()
```
It looks like we can simplify to `self.flags.contains(SymbolFlags::Export) || (self.is_in_ts_namespace() && self.in_export_node())` ?

DonIsaac replied:
Potentially, depending on how well symbol flags are working now. That was the main blocker.

DonIsaac replied:
The module record check is needed for scenarios like this:
```ts
const foo = 1
export { foo }
```

Dunqing commented:
Overall it looks good! BTW what do you think about which part is the slowest in this rule? I would like to deep dive in it!

DonIsaac commented:
> Overall it looks good! BTW what do you think about which part is the slowest in this rule? I would like to deep dive in it!

If I had to guess, I'd say the following are the most expensive:
1. ignore checks, but only if any ignore patterns are set. They rely on regular expressions.
2. Usage checks. For logical isolation, each check kind is separated into a separate function, but each requires a traversal up the AST. Sometimes these checks walk all the way up.
3. "allow" checks for function arguments. They require usage checks on all following parameters.

DonIsaac commented:
This PR is now ready to be merged, once everyone is OK with it.

mysteryven commented:
LGTM, this is impressive work!!

---

Found several cases should fail :)

1. https://github.com/preactjs/preact/blob/f7f9d9bf4309d09bcd0a432acd72787b7d9baecb/compat/src/index.d.ts#L674-L680
```js
// SS is unused
export abstract class PureComponent<P = {}, S = {}, SS = any> extends preact.Component<P, S> {
    isPureReactComponent: boolean;
}
```

2. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/87ffcec0275bcded1270f737deb0bc675129a7f6/types/angular-material/index.d.ts#L1

```js
import * as angular from "angular";
```

3. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/87ffcec0275bcded1270f737deb0bc675129a7f6/types/antlr4/CommonTokenStream.d.ts#L3-L4

4. Looks like some test files are not lint, it might not the issue about this PR, they are in `diff.txt`: https://stackblitz.com/edit/stackblitz-starters-7d4upf?file=diff.txt.

Dunqing commented:
> @mysteryven those examples are located in `.d.ts` files, which I am currently skipping checks on completely.

You are right. The semantic doesn't handle `.d.ts` files either currently.

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/mod.rs:
```diff
@@ -0,0 +1,320 @@
+
+impl NoUnusedVars {
+    fn run_on_symbol_internal<'a>(&self, symbol: &Symbol<'_, 'a>, ctx: &LintContext<'a>) {
+        // println!("{:#?}", symbol);
```
```suggestion
```

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/binding_pattern.rs:
```diff
@@ -0,0 +1,68 @@
+pub(super) struct BindingContext<'s, 'a> {
+    pub options: &'s NoUnusedVars,
+    pub semantic: &'s Semantic<'a>,
+    // pub symbol: &'s Symbol<'s, 'a>,
```
```suggestion
```

Dunqing commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/symbol.rs:
```diff
@@ -0,0 +1,225 @@
+        let Some(reference_id) = other.reference_id.get() else {
+            return false;
+        };
+        // TODO: which is cheaper; a symbol_id check or a name check?
+        // if self.name() == other.name {
+        //     return true;
+        // }
```
```suggestion
```

We must use `symbol_id` checking. The `IdentifierReference` of the same name in different scopes may not be equal.

Dunqing reviewed (approved):
Awesome! Let's merge it after all requested changes are resolved!

DonIsaac commented:
@Dunqing I think we're good to go, should we send it?

DonIsaac committed:
feat(linter): add eslint/no-unused-vars (⭐ attempt 3.2) (#4445)

> Re-creation of #4427 due to rebasing issues. Original attempt: #642
-----

Third time's the charm?

Each time I attempt this rule, I find a bunch of bugs in `Semantic`, and I expect this attempt to be no different. Expect sidecar issues+PRs stemming from this PR here.

## Not Supported
These are cases supported in the original eslint rule, but that I'm intentionally deciding not to support
- export comments in scripts
  ```js
  /* exported a */ var a;
  ```
- global comments
  ```js
  /* global a */ var a;
   ```

## Behavior Changes
These are intentional deviations from the original rule's behavior:
- logical re-assignments are not considered usages
  ```js
  // passes in eslint/no-unused-vars, fails in this implementation
  let a = 0; a ||= 1;
  let b = 0; b &&= 2;
  let c = undefined; c ??= []
  ```

## Known Limitations
- Lint rules do not have babel or tsconfig information, meaning we can't determine if `React` imports are being used or not. The relevant tsconfig settings here are `jsx`, `jsxPragma`, and `jsxFragmentName`. To accommodate this, all imports to symbols named `React` or `h` are ignored in JSX files.
- References to symbols used in JSDoc `{@link}` tags are not created, so symbols that are only used in doc comments will be reported as unused. See: #4443
- `.vue` files are skipped completely, since variables can be used in templates in ways we cannot detect
  > note: `.d.ts` files are skipped as well.

## Todo
- [x] Skip unused TS enum members on used enums
- [x] Skip unused parameters followed by used variables in object/array spreads
- [x] Re-assignments to array/object spreads do not respect `destructuredArrayIgnorePattern` (related to: https://github.com/oxc-project/oxc/issues/4435)
- [x] #4493
- [x] References inside a nested scope are not considered usages (#4447)
- [x] Port over typescript-eslint test cases _(wip, they've been copied and I'm slowly enabling them)_
- [x] Handle constructor properties
  ```ts
  class Foo {
    constructor(public a) {} // `a` should be allowed
  }
  ```
- [x] Read references in sequence expressions (that are not in the last position) should not count as a usage
  ```js
  let a = 0; let b = (a++, 0); console.log(b)
  ```
  > Honestly, is anyone even writing code like this?
- [x] function overload signatures should not be reported
- [x] Named functions returned from other functions get incorrectly reported as unused (found by @camc314)
  ```js
  function foo() {
    return function bar() { }
  }
  Foo()()
  ```
- [x] false positive for TS modules within ambient modules
  ```ts
  declare global {
    // incorrectly marked as unused
    namespace jest { }
  }
  ```

## Blockers
- https://github.com/oxc-project/oxc/issues/4436
- https://github.com/oxc-project/oxc/issues/4437
- #4446
- #4447
- #4494
- #4495

## Non-Blocking Issues
- #4443
- #4475 (prevents checks on exported symbols from namespaces)

Dunqing commented:
@DonIsaac  Can you take a look at https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/10173564561/job/28137965651. This case seems a false positive

```shell
  x eslint(no-unused-vars): Function 'onMessage' is declared but never used.
    ,-[wasm-runtime/fs-proxy.cjs:78:51]
 77 |  */
 78 | module.exports.createOnMessage = (fs) => function onMessage(e) {
    :                                                   ^^^^|^^^^
    :                                                       `-- 'onMessage' is declared here
 79 |   if (e.data.__fs__) {
    `----
  help: Consider removing this declaration.
```

armano2 commented:
> @DonIsaac Can you take a look at https://github.com/oxc-project/oxlint-ecosystem-ci/actions/runs/10173564561/job/28137965651. This case seems a false positive
> 
> ```shell
>   x eslint(no-unused-vars): Function 'onMessage' is declared but never used.
>     ,-[wasm-runtime/fs-proxy.cjs:78:51]
>  77 |  */
>  78 | module.exports.createOnMessage = (fs) => function onMessage(e) {
>     :                                                   ^^^^|^^^^
>     :                                                       `-- 'onMessage' is declared here
>  79 |   if (e.data.__fs__) {
>     `----
>   help: Consider removing this declaration.
> ```

this is indeed a bug, it looks that `argsIgnorePattern` is also not supported
